### PR TITLE
Compress Data in EntityStore

### PR DESCRIPTION
**Pull request by tubbo**
Using the `:compress` option, compress and decompress data over the wire
to Redis. This improves bandwidth consumption, RAM usage, and reduces
the amount of storage space needed to run your Redis cache server.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -5,3 +5,5 @@ git_source :github do |name|
   repo = name =~ %r{/} ? name : "#{name}/#{name}"
   "https://github.com/#{repo}.git"
 end
+
+gem 'snappy'
```
Comment (tubbo) [inline Gemfile]: not an additional dependency, just allows testing of custom compressors.

File: test/rack/cache/entitystore/redis_test.rb [hunk 2]
```diff
@@ -18,7 +57,7 @@ def sha_like?
 
   it 'properly delegates the TTL to redis' do
     @store = ::Rack::Cache::EntityStore::Redis.new({ :host => 'localhost' }, { :default_ttl => 120 })
-    key, size = @store.write(['She rode to the devil,'])
+    key, _size = @store.write(['She rode to the devil,'])
```
Comment (tubbo) [inline test/rack/cache/entitystore/redis_test.rb]: all these var renames are just rubocop fixes.

File: lib/rack/cache/redis_entitystore.rb [hunk 3]
```diff
@@ -1,34 +1,93 @@
...
-          cache.get key
+          decompress cache.get(key)
         end
 
         def write(body, ttl=0)
           buf = StringIO.new
-          key, size = slurp(body) {|part| buf.write(part) }
+          key, _size = slurp(body) { |part| buf.write(part) }
           ttl = ttl.to_i.zero? ? default_ttl : ttl
+          value = compress decompress(buf.string)
```
Comment (tubbo) [inline lib/rack/cache/redis_entitystore.rb]: why did i have to do this? who knows! but it was the key to getting shit to work properly...

**Comment by tubbo** [comment]
>     1. Why did you implement compression only for this gem? Is this approach applicable to `redis-store`? In this way, the entire set of gems can benefit from it.

I don't believe it's actually necessary in the main gem for a few reasons. For starters, the only other place I can think of where the size of a cache fragment would actually get to the point of needing compression would be redis-activesupport, which already benefits from ActiveSupport::Cache::Entry's use of compression on objects above 1K. I took a similar approach in this gem.

>     2. Assuming very low latency between the app server and Redis server, is compression worth? Compression and decompression will have their own downsides like memory allocation, CPU consumption, GC pressure. Did you run benchmarks to understand the effects of compression on the app that uses it, and what's the memory savings for Redis?

Compression is possibly _not_ worth it, which is why it's off by default. The reason why I built this was because an application we maintain at work was hitting the upper limit of AWS bandwidth allocation between ElastiCache and the app server(s), which apparently is 15Gbps. I opened this pull request primarily to show you and the rest of the core team what we were doing, but it's also where we can collect test results and benchmarks to make sure this is truly a good idea.

>     3. Would this require a major release bump (assuming SemVer)? If I have an app running this gem, I turn on compression, am I still able to read/write already existing data from Redis? Or what if I want to switch adapter, from builtin solution to `snappy`? Should we think of a migration tool?

I was trying to make it relatively painless to turn on compression, so when you read something from Redis that isn't compressed, you won't get an error. That said, switching compression algorithms at the moment might be a little problematic, you'd have to blow away the cache in order for it to work properly.

**Comment by jodosha** [comment]
@tubbo thanks, your comments make sense.

I'll try to review the code soon. To be clear, is this PR an extraction of already existing code used in your production app?

Anyone else in @redis-store/core-team has an opinion?

**Comment by jodosha** [comment]
@tubbo Gotcha. Let us to know how it goes. Please make sure to give credit to the original code. 👍

**Comment by tubbo** [comment]
Initial testing proved that a lot of assumptions I had about Zlib and Rack::Cache were fundamentally wrong 🤣. Dealt with a lot of garbled text and incomplete responses until my teammate @MMartyn discovered that the encoding coming out of `Zlib::Inflate` was actually ASCII, not UTF-8. Additionally, it turns out that you can't pass options down through Rack::Cache, which I'm probably going to open a proposal PR on that project about. Right now, the option is configured by setting a value on the `Redis::Rack::Cache` module, which isn't ideal but it's the only way I can think of getting this to work without additional code in rack-cache.

**Comment by tubbo** [comment]
### Quick update on this...

We were able to do some real-world load testing on a staging server and found that, while overall time is pretty much the same due to the application needing to compress/decompress the data as it comes over the wire, throughput to the Redis server was cut by almost 80%. We did this load test yesterday, November 7th, 2019.

As you can see, the load for `Rack::Cache` in the application pretty much stays the same throughout the last seven days:

Screen Shot 2019-11-08 at 12 01 28 PM

However, during the last load test (prior to enabling compression), this is what the traffic to Redis looked like:

Screen Shot 2019-11-08 at 11 55 07 AM

And now, after compression:

Screen Shot 2019-11-08 at 11 55 36 AM

As you can see, the same amount of load applied to the application is sending almost 80% less traffic to Redis.

### So what does this mean?

This feature is certainly not something that every application is going to need, which is one reason why it will remain off by default. But if you're maintaining an application that's experiencing an extreme level of load at certain times of the day, so much so that your network traffic to the Redis server becomes the bottleneck, this will alleviate some of that pain before you have to jump into configuring a Redis cluster or distributed servers.

I'm going to merge this because it seems to have little effect on the application's memory and CPU usage, and provides a benefit for large-scale applications that rely on HTTP cache reads to maintain performance.

**Comment by tubbo** [comment]
### Quick update on this...

We were able to do some real-world load testing on a staging server and found that, while overall time is pretty much the same due to the application needing to compress/decompress the data as it comes over the wire, throughput to the Redis server was cut by almost 80%. We did this load test yesterday, November 7th, 2019. Compression was enabled on November 3rd, 2019.

Over the last seven days, the load for `Rack::Cache` in the application pretty much stays the same:

Screen Shot 2019-11-08 at 12 01 28 PM

You can see when compression was enabled on the 3rd, as the get time decreases and the middleware time increases.

However, during the last load test (prior to enabling compression), this is what the traffic from Redis to the app servers looked like:

Screen Shot 2019-11-08 at 11 55 07 AM

And now, after compression:

Screen Shot 2019-11-08 at 11 55 36 AM

As you can see, Redis is now sending **approximately 80% less data** back to the application over the wire, under the same amount of load stress.

### So what does this mean?

This feature is certainly not something that every application is going to need, which is one reason why it will stay off by default. But if you're maintaining an application that's experiencing an extreme level of load at certain times of the day, so much so that your network traffic to the Redis server becomes the bottleneck, this will alleviate some of that pain before you have to jump into configuring a Redis cluster or distributed servers.

I'm going to merge this because it seems to have little effect on the application's memory and CPU usage, and provides a benefit for large-scale applications that rely on HTTP cache reads to maintain performance.

**Comment by jodosha** [comment]
@tubbo Thanks for taking care of this and for the follow up from the production standpoint. My apologize if I didn't provide a review in time. Please let us to know how it goes in the long run. Cheers!

**Comment by tubbo** [comment]
No problem. Seems to be going well for the site we deployed it to. :)

**Pull request merged**